### PR TITLE
Jardilio route redirect fix

### PR DESCRIPTION
**Pull request by petebacondarwin**
See #3332

### CODE REVIEW

**Comment by petebacondarwin** [comment]
I accidentally changed the authorship of the first commit. I will change back before merging...

File: test/ngRoute/routeSpec.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 'use strict';
 
-describe('$route', function() {
+fdescribe('$route', function() {
```
Comment (gkalpak) [inline test/ngRoute/routeSpec.js]: :scream:

File: src/ngRoute/route.js [hunk 2]
```diff
@@ -588,46 +588,25 @@ function $RouteProvider() {
...
+                       .search(nextRoute.params)
                        .replace();
+              newUrl = $location.url();
             } else {
-              $location.url(nextRoute.redirectTo(nextRoute.pathParams, $location.path(), $location.search()))
-                       .replace();
+              newUrl = nextRoute.redirectTo(nextRoute.pathParams, $location.path(), $location.search());
+              $location.url(newUrl).replace();
+            }
+            if (isDefined(newUrl) && url !== newUrl) {
```
Comment (gkalpak) [inline src/ngRoute/route.js]: I thought `null` would behave the same as `undefined`.

**Comment by gkalpak** [comment]
Where are the docs ? :stuck_out_tongue:

**Comment by gkalpak** [comment]
I would also mention `resolve` amoong the things that don't get processed in the BC notice.

File: src/ngRoute/route.js (hunk 2 above)
Comment (Narretz) [inline src/ngRoute/route.js]: Should all falsy values abort the redirect?

File: test/ngRoute/routeSpec.js [hunk 3]
```diff
@@ -1061,6 +1061,55 @@ describe('$route', function() {
             .toEqual(['http://server/#!/bar/id3?extra=eId', true, null]);
       });
     });
+
+    it('should not instantiate controller or process template for a redirected route', function() {
```
Comment (Narretz) [inline test/ngRoute/routeSpec.js]: What about `resolve`?

**Comment by petebacondarwin** [comment]
So at the moment null works differently so this pr does not change that. We should consider making null work like undefined in a follow up pr

**Comment by petebacondarwin** [comment]
I'll add checks commit message and docs for resolve

File: src/ngRoute/route.js (hunk 2 above)
Comment (petebacondarwin) [inline src/ngRoute/route.js]: currently `null` and `""` are not treated like `undefined` and so this PR doesn't change that (to minimize BC at this stage).

File: test/ngRoute/routeSpec.js (hunk 1 above)
Comment (petebacondarwin) [inline test/ngRoute/routeSpec.js]: Oops

**Comment by petebacondarwin** [comment]
@gkalpak and @Narretz PTAL!

**Comment by Narretz** [comment]
LGTM

**Comment by Narretz** [comment]
Don't forget to change back the authorship though!

File: src/ngRoute/route.js [hunk 4]
```diff
@@ -136,6 +136,12 @@ function $RouteProvider() {
    *      The custom `redirectTo` function is expected to return a string which will be used
    *      to update `$location.path()` and `$location.search()`.
    *
+   *      Routes that specify `redirectTo` will not have their controllers, template functions
+   *      or resolves called, the `$location` will be changed to the redirect url and route
+   *      processing will stop. The exception to this is if the `redirectTo` is a function that
+   *      returns `undefined`. In this case the route transition occurs as though their was no
```
Comment (gkalpak) [inline src/ngRoute/route.js]: _their_  -->  _there_

Comment (petebacondarwin) [inline src/ngRoute/route.js]: Aghast!

**Comment by gkalpak** [comment]
The commit message mentions that `null` cancels the redirection. Remove it when merging to avoid confusion.

LGTM

**Comment by petebacondarwin** [comment]
Oops. Well I guess I have to fix the `null` path now then, eh?

**Pull request closed without merging**